### PR TITLE
feat: Bank remittance report for tracking payroll-entry transactions

### PR DESCRIPTION
**Pull request by Anurag810**
Screenshot 2019-03-25 at 6 09 26 PM

### CODE REVIEW

File: erpnext/hr/report/bank_remittance/bank_remittance.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+def get_salary_slips(payroll_entries):
+	d = list(map(lambda x: {k:v for k, v in x.items() if k == 'name'}, payroll_entries))
+	payroll = ', '.join(map(str, [l['name'] for l in d if 'name' in l]))
+	salary_slips = get_record("Salary Slip", [("payroll_entry", "IN", payroll)],
+		fields = ["modified", "net_pay", "bank_name", "bank_account_no", "payroll_entry", "employee", "employee_name", "status"]
+	)
+
+	# appending company debit accounts
+	for slip in salary_slips:
+		for entry in payroll_entries:
```
Comment (nabinhait) [inline erpnext/hr/report/bank_remittance/bank_remittance.py]: This is unnecessarily complicated. No need to write nested for loop.

File: erpnext/hr/report/bank_remittance/bank_remittance.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+				slip["debit_acc_no"] = entry.company_account
+
+	return salary_slips
+
+def get_emp_bank_ifsc_code(salary_slips):
+	d = list(map(lambda x: {k:v for k, v in x.items() if k == 'employee'}, salary_slips))
+	emp_names = ', '.join(map(str, [l['employee'] for l in d if 'employee' in l]))
+	ifsc_codes = get_record("Employee", [("name", "IN", emp_names)], ["ifsc_code", "name"])
+	for slip in salary_slips:
+		for code in ifsc_codes:
```
Comment (nabinhait) [inline erpnext/hr/report/bank_remittance/bank_remittance.py]: Nested for loop can be avoided

File: erpnext/hr/report/bank_remittance/bank_remittance.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+			if code.name == slip.employee:
+				slip["ifsc_code"] = code.ifsc_code
+
+	return salary_slips
+
+def get_company_account(payment_accounts ,payroll_entries):
+	company_accounts = get_record("Bank Account", [("account", "IN", payment_accounts)], ["account", "bank_account_no"])
+	for acc in company_accounts:
+		for entry in payroll_entries:
```
Comment (nabinhait) [inline erpnext/hr/report/bank_remittance/bank_remittance.py]: Nested for loop can be avoided

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/hr/report/bank_remittance/bank_remittance.py (hunk 1 above)
Comment (Anurag810) [reply erpnext/hr/report/bank_remittance/bank_remittance.py]: done

File: erpnext/hr/report/bank_remittance/bank_remittance.py (hunk 2 above)
Comment (Anurag810) [reply erpnext/hr/report/bank_remittance/bank_remittance.py]: done

File: erpnext/hr/report/bank_remittance/bank_remittance.py (hunk 3 above)
Comment (Anurag810) [reply erpnext/hr/report/bank_remittance/bank_remittance.py]: done

File: erpnext/hr/report/bank_remittance/bank_remittance.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	for acc in company_accounts:
+		company_accounts_map[acc.account] = acc
+
+	for entry in payroll_entries:
+		entry["company_account"] = company_accounts_map[entry.payment_account]['bank_account_no']
+
+	return payroll_entries
+
+def get_record(doctype, filter, fields):
+	return frappe.get_all(doctype, filters=filter, fields=fields)
```
Comment (netchampfaris) [inline erpnext/hr/report/bank_remittance/bank_remittance.py]: Why not use `get_all` directly, this is not needed.

File: erpnext/hr/report/bank_remittance/bank_remittance.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	accounts = frappe.get_all("Account",
+	filters={
+		"account_type": "Bank"
+	}, as_list =1)
+	return accounts
+
+def get_payroll_entry(accounts, filters):
+	accounts = list(itertools.chain(*accounts))
+	payroll_filter = [
+			('payment_account', 'IN', accounts),
```
Comment (netchampfaris) [inline erpnext/hr/report/bank_remittance/bank_remittance.py]: indenting

File: erpnext/hr/report/bank_remittance/bank_remittance.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+					"currency": frappe.get_cached_value('Company', filters.company, 'default_currency'),
+					"amount": salary.net_pay,
+				}
+
+			data.append(row)
+	return columns, data
+
+def get_account():
+	accounts = frappe.get_all("Account",
+	filters={
```
Comment (netchampfaris) [inline erpnext/hr/report/bank_remittance/bank_remittance.py]: Indentation

File: erpnext/hr/report/bank_remittance/bank_remittance.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		"account_type": "Bank"
+	}, as_list =1)
```
Comment (netchampfaris) [inline erpnext/hr/report/bank_remittance/bank_remittance.py]:
```suggestion
	}, as_list=1)
```

File: erpnext/hr/report/bank_remittance/bank_remittance.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+	accounts = get_account()
+	payroll_entries = get_payroll_entry(accounts, filters)
+	salary_slips = get_salary_slips(payroll_entries)
+	get_emp_bank_ifsc_code(salary_slips)
+
+	for salary in salary_slips:
+		if salary.bank_name and salary.bank_account_no and salary.debit_acc_no and salary.status in ["Submitted", "Paid"]:
+			row = {
+					"payroll_no": salary.payroll_entry,
```
Comment (netchampfaris) [inline erpnext/hr/report/bank_remittance/bank_remittance.py]: Indentation

**Review by netchampfaris — CHANGES_REQUESTED** [review]

File: erpnext/hr/report/bank_remittance/bank_remittance.py (hunk 8 above)
Comment (Anurag810) [reply erpnext/hr/report/bank_remittance/bank_remittance.py]: done

File: erpnext/hr/report/bank_remittance/bank_remittance.py (hunk 5 above)
Comment (Anurag810) [reply erpnext/hr/report/bank_remittance/bank_remittance.py]: done

File: erpnext/hr/report/bank_remittance/bank_remittance.py (hunk 6 above)
Comment (Anurag810) [reply erpnext/hr/report/bank_remittance/bank_remittance.py]: done

File: erpnext/hr/report/bank_remittance/bank_remittance.py (hunk 7 above)
Comment (Anurag810) [reply erpnext/hr/report/bank_remittance/bank_remittance.py]: done

File: erpnext/hr/report/bank_remittance/bank_remittance.py (hunk 4 above)
Comment (Anurag810) [reply erpnext/hr/report/bank_remittance/bank_remittance.py]: done

**Review by netchampfaris — APPROVED** [review]

**Pull request merged**